### PR TITLE
ENH: Add ignore_index to sort_index

### LINKED ISSUE #30114: API: add ignore_index keyword to .sort_* & .drop_duplicates
I see lots of code that goes like this

```
In [17]: df = pd.DataFrame({'A': [1, 2, 3]})                                                                                                                                        

In [18]: df                                                                                                                                                                         
Out[18]: 
   A
0  1
1  2
2  3

In [19]: df.sort_values('A', ascending=False)                                                                                                                                       
Out[19]: 
   A
2  3
1  2
0  1

In [20]: df.sort_values('A', ascending=False).reset_index(drop=True)                                                                                                                
Out[20]: 
   A
0  3
1  2
2  1
```

might be nice from an API / consistency perspective to add a ``ignore_index=False|True`` keyword to ``.sort_values()``, and ``.drop_duplicates()`` that does the reset in-line; this would give consistency similar to a ``pd.concat([......], ignore_index=True)`` operation which users are very familiar

### PR DESCRIPTION
**Pull request by charlesdong1991**
- [x] closes #30114
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

@jreback 
I just looked back at #30114 and found out that i overlooked to add `ignore_index` to `sort_index`, and this might be the reason this issue is still left open. I added this to `sort_index` to close this issue.

### CODE REVIEW

File: pandas/tests/series/methods/test_sort_index.py [hunk 1]
```diff
@@ -135,3 +135,35 @@ def test_sort_index_intervals(self):
...
+    )
+    def test_sort_index_ignore_index(
+        self, original_list, sorted_list, ascending, ignore_index, output_index
+    ):
+        # GH 30114
+        # GH 30114
+        sr = Series(original_list)
+        expected = Series(sorted_list, index=output_index)
+
+        # Test when inplace is False
```
Comment (simonjayhawkins) [inline pandas/tests/series/methods/test_sort_index.py]: can inplace be parameterised easily

File: pandas/core/frame.py [hunk 2]
```diff
@@ -4827,6 +4827,7 @@ def sort_index(
         kind="quicksort",
         na_position="last",
         sort_remaining=True,
+        ignore_index=False,
```
Comment (alimcmaster1) [inline pandas/core/frame.py]: Can we type any new args?

File: pandas/tests/series/methods/test_sort_index.py [hunk 3]
```diff
@@ -135,3 +135,35 @@ def test_sort_index_intervals(self):
...
+            ([2, 3, 6, 1], [1, 6, 3, 2], False, True, [0, 1, 2, 3]),
+            ([2, 3, 6, 1], [1, 6, 3, 2], False, False, [3, 2, 1, 0]),
+        ],
+    )
+    def test_sort_index_ignore_index(
+        self, original_list, sorted_list, ascending, ignore_index, output_index
+    ):
+        # GH 30114
+        # GH 30114
+        sr = Series(original_list)
```
Comment (simonjayhawkins) [inline pandas/tests/series/methods/test_sort_index.py]: not seen sr used, s or ser?

File: pandas/core/series.py [hunk 4]
```diff
@@ -2880,6 +2880,7 @@ def sort_index(
         kind="quicksort",
         na_position="last",
         sort_remaining=True,
+        ignore_index=False,
```
Comment (simonjayhawkins) [inline pandas/core/series.py]: can you add type annotations when adding kwargs

File: pandas/core/frame.py (hunk 2 above)
Comment (simonjayhawkins) [inline pandas/core/frame.py]: type annotation

**Review by simonjayhawkins — COMMENTED** [review]
Thanks @charlesdong1991  for the PR.

File: pandas/core/frame.py (hunk 2 above)
Comment (charlesdong1991) [reply pandas/core/frame.py]: yes, added!

File: pandas/tests/series/methods/test_sort_index.py [hunk 5]
```diff
@@ -135,3 +135,35 @@ def test_sort_index_intervals(self):
...
+            ([2, 3, 6, 1], [2, 3, 6, 1], True, False, [0, 1, 2, 3]),
+            ([2, 3, 6, 1], [1, 6, 3, 2], False, True, [0, 1, 2, 3]),
+            ([2, 3, 6, 1], [1, 6, 3, 2], False, False, [3, 2, 1, 0]),
+        ],
+    )
+    def test_sort_index_ignore_index(
+        self, original_list, sorted_list, ascending, ignore_index, output_index
+    ):
+        # GH 30114
+        # GH 30114
```
[the remark this answers is not in the document]
Comment (charlesdong1991) [reply pandas/tests/series/methods/test_sort_index.py]: oh, sorry, removed!

File: pandas/tests/series/methods/test_sort_index.py (hunk 1 above)
Comment (charlesdong1991) [reply pandas/tests/series/methods/test_sort_index.py]: emm, i think in this case, if using parameterization, might not look very clean

File: pandas/tests/series/methods/test_sort_index.py (hunk 3 above)
Comment (charlesdong1991) [reply pandas/tests/series/methods/test_sort_index.py]: changed to `ser`

File: pandas/core/series.py (hunk 4 above)
Comment (charlesdong1991) [reply pandas/core/series.py]: added!

File: pandas/core/frame.py (hunk 2 above)
Comment (charlesdong1991) [reply pandas/core/frame.py]: added!

File: pandas/tests/frame/methods/test_sort_index.py [hunk 6]
```diff
@@ -229,3 +229,84 @@ def test_sort_index_intervalindex(self):
...
         result = result.columns.levels[1].categories
         tm.assert_index_equal(result, expected)
+
+    @pytest.mark.parametrize(
+        "original_dict, sorted_dict, ascending, ignore_index, output_index",
+        [
+            ({"A": [1, 2, 3]}, {"A": [3, 2, 1]}, False, True, [0, 1, 2]),
+            ({"A": [1, 2, 3]}, {"A": [1, 2, 3]}, True, True, [0, 1, 2]),
+            ({"A": [1, 2, 3]}, {"A": [3, 2, 1]}, False, False, [2, 1, 0]),
+            ({"A": [1, 2, 3]}, {"A": [1, 2, 3]}, True, False, [0, 1, 2]),
```
Comment (WillAyd) [inline pandas/tests/frame/methods/test_sort_index.py]: Minor but can you use a non-default index as part of the expectation? Right now the parametrization can't disambiguate between the default index and `ignore_index` argument

File: pandas/tests/series/methods/test_sort_index.py (hunk 1 above)
Comment (WillAyd) [reply pandas/tests/series/methods/test_sort_index.py]: I think this should be parametrized - what do you think makes it less readable?

**Review by WillAyd — CHANGES_REQUESTED** [review]
Looks generally good - a few comments

File: pandas/core/series.py [hunk 7]
```diff
@@ -118,6 +118,8 @@ def wrapper(self):
 # ----------------------------------------------------------------------
 # Series class
 
+bool_t = bool  # Need alias because NDFrame has def bool:
```
Comment (jreback) [inline pandas/core/series.py]: don't we already have this defined? @simonjayhawkins

File: pandas/tests/series/methods/test_sort_index.py (hunk 1 above)
Comment (jreback) [reply pandas/tests/series/methods/test_sort_index.py]: ok with trying parameterize as a followup, as for .sort_values and .duplicated these tests look very much like this.

File: pandas/tests/frame/methods/test_sort_index.py (hunk 6 above)
Comment (charlesdong1991) [reply pandas/tests/frame/methods/test_sort_index.py]: thanks! changed!

File: pandas/tests/series/methods/test_sort_index.py (hunk 1 above)
Comment (charlesdong1991) [reply pandas/tests/series/methods/test_sort_index.py]: emm, okay! i thought with `inplace` being `Ture`, we do not need to do something like `df = df.sort_index(inplace=True)`, but `False` needs to be like this, and testing object will be different, one is testing if original changes, another is to test if assigned new df is changed. 

Sorry to bring up discussion here, it definitely can be parametrized, was just my preference on such case. I will parametrize all three methods in the follow-up PR for sure!

File: pandas/core/series.py (hunk 7 above)
Comment (simonjayhawkins) [reply pandas/core/series.py]: we have this in pandas.core.generic where bool method is defined. bool is not overridden here (or in core.frame) so don't need alias.

@charlesdong1991 u can remove this

File: pandas/core/frame.py [hunk 8]
```diff
@@ -309,6 +309,8 @@
 # -----------------------------------------------------------------------
 # DataFrame class
 
+bool_t = bool  # Need alias because NDFrame has def bool:
```
Comment (simonjayhawkins) [inline pandas/core/frame.py]: remove

File: pandas/core/frame.py [hunk 9]
```diff
@@ -4831,6 +4833,7 @@ def sort_index(
         kind="quicksort",
         na_position="last",
         sort_remaining=True,
+        ignore_index: bool_t = False,
```
Comment (simonjayhawkins) [inline pandas/core/frame.py]:
```suggestion
        ignore_index: bool = False,
```

File: pandas/core/series.py [hunk 10]
```diff
@@ -2963,6 +2965,7 @@ def sort_index(
         kind="quicksort",
         na_position="last",
         sort_remaining=True,
+        ignore_index: bool_t = False,
```
Comment (simonjayhawkins) [inline pandas/core/series.py]:
```suggestion
        ignore_index: bool = False,
```

**Comment by WillAyd** [comment]
Lgtm

**Review by jreback — APPROVED** [review]
lgtm. if you can followup with a parameterization / deduplication of tests PR would be great

**Comment by charlesdong1991** [comment]
thanks! will do @jreback

**Comment by stonereese** [comment]
Hi,`ignore_index = True` works  just fine and meets my expectations when `axis =0`,but when `axis = 1`, the number of index after reset using `ignore = Ture` is equal to the number of columns instead of the original index number,like `new_index = range(len(columns))`.
Is this a bug, or where am I going wrong?
THKS!

**Pull request merged**